### PR TITLE
Example mlp

### PR DESCRIPTION
**Pull request by aaronkl**
replaces #687

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: examples/hnas/example_mlp.py
Comment (mseeger) [inline examples/hnas/example_mlp.py:26]: You can drop the **kwargs. I changed the promotion-based HB not to supply resume_from this way anymore.

File: examples/hnas/example_mlp.py
Comment (mseeger) [inline examples/hnas/example_mlp.py:38]: Can you give a more precise instruction of what the user has to do?

File: examples/hnas/example_mlp.py
Comment (mseeger) [inline examples/hnas/example_mlp.py:102]: Don't we get this anyway in training_history?

File: examples/hnas/example_mlp.py
Comment (mseeger) [inline examples/hnas/example_mlp.py:175]: Can you check examples/hpo/mlp.py? We have quite a number of other args the user may want to change here.

File: examples/hnas/example_mlp.py
Comment (mseeger) [inline examples/hnas/example_mlp.py:159]: One of num_trials, timeout is allowed to be None

File: examples/hnas/example_mlp.py
Comment (mseeger) [inline examples/hnas/example_mlp.py:221]: User argument?

File: examples/hnas/example_mlp.py
Comment (mseeger) [inline examples/hnas/example_mlp.py:225]: User argument?

File: examples/hnas/example_mlp.py
Comment (mseeger) [inline examples/hnas/example_mlp.py:228]: Maybe call callback one more time here, to be sure?

**Comment by szha** [comment]
Job PR-765-1 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-765/1/index.html

File: examples/hnas/example_mlp.py
Comment (jwmueller) [inline examples/hnas/example_mlp.py:1]: Add comment at the top of this file describing what this example is supposed to show. Also consider explicitly stating what auxiliary packages must be installed to run this example.

File: README.md
Comment (jwmueller) [inline README.md:87]: Use more informative name here

File: examples/hnas/example_mlp.py
Comment (jwmueller) [inline examples/hnas/example_mlp.py:157]: Don't think users will understand what `autoff benchmarks` is. May want to describe it more or provide external link to more info.

File: examples/hnas/example_mlp.py
Comment (aaronkl) [reply examples/hnas/example_mlp.py:26]: ok dropped

File: examples/hnas/example_mlp.py
Comment (aaronkl) [reply examples/hnas/example_mlp.py:38]: ok I added some python commands to load this data

File: examples/hnas/example_mlp.py
Comment (aaronkl) [reply examples/hnas/example_mlp.py:102]: No I don't think that the configuration is stored in the training history. But maybe we should add this.

File: examples/hnas/example_mlp.py
Comment (aaronkl) [reply examples/hnas/example_mlp.py:175]: ok I've added a few more

File: examples/hnas/example_mlp.py
Comment (aaronkl) [reply examples/hnas/example_mlp.py:159]: ok I add a brief description

File: examples/hnas/example_mlp.py
Comment (aaronkl) [reply examples/hnas/example_mlp.py:225]: I am not exactly sure what you mean here

File: examples/hnas/example_mlp.py
Comment (aaronkl) [reply examples/hnas/example_mlp.py:228]: I added a line that shows how to access the final training_history

File: examples/hnas/example_mlp.py
Comment (aaronkl) [reply examples/hnas/example_mlp.py:1]: ok added

File: README.md
Comment (aaronkl) [reply README.md:87]: added

File: examples/hnas/example_mlp.py
Comment (aaronkl) [reply examples/hnas/example_mlp.py:157]: good catch, was an artifact

File: examples/hnas/example_mlp.py
Comment (mseeger) [reply examples/hnas/example_mlp.py]: I mean we could make grace_period and reduction_factor arguments in the CommandParser.

**Comment by szha** [comment]
Job PR-765-2 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-765/2/index.html

**Review by mseeger — APPROVED** [review]

File: examples/hnas/example_mlp.py
Comment (mseeger) [inline examples/hnas/example_mlp.py:255]: typo

File: examples/hnas/example_mlp.py
Comment (mseeger) [inline examples/hnas/example_mlp.py:232]: Remove

**Review by mseeger — APPROVED** [review]

File: examples/hnas/example_mlp.py
Comment (jwmueller) [inline examples/hnas/example_mlp.py:12]: Consider linking this pytorch tutorial as well: https://auto.gluon.ai/stable/tutorials/torch/hpo.html

and briefly describing what the differences are. 

Why are both tutorials useful? It appears to me a key difference is the emphasis on hyperband-style HPO in this example, in which case the comment at the top should probably make this clear.

File: README.md
Comment (jwmueller) [inline README.md:87]: I think this could be more clearly written as something like:

"Example script for efficient Hyperband-style tuning of PyTorch neural network hyperparameters"

File: examples/hnas/example_mlp.py
Comment (aaronkl) [reply examples/hnas/example_mlp.py]: makes sense, I've added that

File: examples/hnas/example_mlp.py
Comment (aaronkl) [reply examples/hnas/example_mlp.py:255]: fixed

File: examples/hnas/example_mlp.py
Comment (aaronkl) [reply examples/hnas/example_mlp.py:232]: done

File: examples/hnas/example_mlp.py
Comment (aaronkl) [reply examples/hnas/example_mlp.py:12]: you're right, the main focus here is successive-halving based HPO. I made this more clear

File: README.md
Comment (aaronkl) [reply README.md:87]: ok I changed it to "Example scripts for efficient multi-fidelity HNAS of PyTorch neural network model" (more scripts to follow)

File: examples/hnas/example_mlp.py
Comment (mseeger) [reply examples/hnas/example_mlp.py]: You could add these commands to download MNIST to the main function, this would then make sure the example runs on a single instance. The tricky part is if somebody runs this on multiple instances.

**Comment by szha** [comment]
Job PR-765-3 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-765/3/index.html

File: examples/hnas/example_mlp.py
Comment (mseeger) [inline examples/hnas/example_mlp.py:226]: This did not produce any log output for me. I had to use the line

logging.basicConfig(level=logging.INFO)

File: examples/hnas/example_mlp.py
Comment (mseeger) [reply examples/hnas/example_mlp.py:25]: We could link this example in the Pytorch tutorial

File: examples/hnas/example_mlp.py
Comment (aaronkl) [reply examples/hnas/example_mlp.py:226]: ok I changed it

**Comment by szha** [comment]
Job PR-765-4 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-765/4/index.html

**Comment by szha** [comment]
Job PR-765-7 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-765/7/index.html

**Comment by Innixma** [comment]
@aaronkl What is the status of this PR?

**Comment by aaronkl** [comment]
from my side it is good to go, just waiting for final approval

**Comment by Innixma** [comment]
Do you mean final approval by Matthias? It looks like he already approved the PR.

**Pull request merged**
